### ISSUE TITLE
Add user groups for finer grained access control to services/modules

### ISSUE THREAD

**Issue by JoKeyser** [freedombox/Plinth#860]
For many use cases, it may not be a good idea that every Freedombox user can access all services. A solution would be to add user groups with restricted access rights. For example users within an 'ejabberd' group can only access the XMPP server ejabberd... or users in a 'chat' group might have access to both ejabberd and Matrix.
@mray and @SunilMohanAdapa may have more to say about this, I mainly opened this issue to start documentation.

**Comment by mray** [comment]
I'm not sure about possible technical solutions, but there is a SSO scenario I want to avoid:
1. User sets up FreedomBox with full SSO support and installs multiple apps.
2. There is only *one username* that only corresponds to *one password*.
3. All apps share the same password (jabber, calendar, wiki, ....).

If we use permissions and groups - what would a default scenario look like?
Creating different users for different apps is a bad workaround, we would use it to map users and apps via one concept.

To me it feels like we are suffering the problem of not being in a position to truly integrate upstream projects, and we probably never will. A solution could be to have an app that does "book keeping" like a password manager and tracks down and notes app credentials for the user. That way a user could login to plinth and look up credentials to set up respective client software.

**Comment by JoKeyser** [comment]
Hi @mray , thanks for adding.
I'm not sure whether I understand correctly... You seem to would like to have the same username for different services, but with different passwords? If so, this would have significant drawbacks: You'd have to remember multiple passwords, probably leading to weaker ones. And in case your client device gets stolen/lost, you also would have to reset all the passwords, not just one.

I guess I don't quite get which threat model you'd like to address. If it's the vulnerability of knowing one App's username/password, would it not be even safer to also have a different username, not just a different password?
On the other hand, if it's more for convenience, then having a single login (with a strong password) would be the ultimate solution in my opinion. I think this is what our target audience would mostly opt for (and frankly, me too).

In your last paragraph, the password manager - would that run on the Freedombox, or on the client side?

**Comment by mray** [comment]
Let me clarify:

### Threat model:
Your assumption about the threat model is correct: I don't want leaked credentials of any app being equivalent to leaked access to ALL freedombox services. We offer third party software and would trust the entire security on the weakest chain (also including devices and clients we have no control over!).

### Same user different password
Multiple passwords, same username makes sense to me as the username is often an identifier on the respective service, not just credential. 

### Drawbacks
You are also right about the drawbacks: A new task of keeping track of credentials to the services.
But that's what FreedomBox should do for you.

### Solution
FreedomBox should help doing security right. I can see how it can gathers all login credentials in an extra module like this:

# Your credentials:
|  | *Server url* | *User*  | *Password* | 
| ------------- |-------------| -----| -----|
| **jabber** | https://jabber.domain.com | username | *5qb45243tvw* |
| **wiki** | https://domain.com/wiki | username | *wbhtvdthrth* |
| **transmission** | https://domain.com/bittorrent | username | *trdzude684e68* |
| **syncthing** | https://domain.com/bittorrent | username | *w5q36wqpihqq346* |

...or show relevant parts in an app card tab. This would make setting up clients generally easier, too.

**Comment by JoKeyser** [comment]
Thanks for the clarification.
I think I get the idea, but as far as I understand it's equivalent to the current implementation and future plans with SSO. Let me explain, and please correct me if I'm wrong.
The current idea is to integrate all apps with LDAP, thus every app relays its login-request to that system. In that sense, there isn't really a "weakest link" (weak implementation) that has access to the same credentials; it's all one single system that we have to trust.
In other words, LDAP already sits in the same role as your suggested extra "master key store" module that uses different internal passwords. Whether LDAP or the master-key falls over, it's the same catastrophic result, thus we must trust a single system in both cases. On the client, only one password is stored in both cases, so I think what you describe is already in place (or on the way).

I also just realised that this is almost the same discussion as in issue #690.

**Comment by mray** [comment]
Disclaimer: I'm almost completely ignorant about what LDAP does. What you said sounds right, but seems to not address my concern. I'm specifically concerned about the credentials I have to repeatedly use with each **client** application.

An example:
* I'm the user "jon" and access plinth with my password "123456"
* I install *radicale* on freedombox
* I configure *davdoid* on my mobile device using "jon" & "123456"
* I install *ejabberd* on freedombox
* I configure *conversations* on my mobile device "jon" & "123456" **again**
* ... *any other* application would use "jon" & "123456" ....

I'm *not* concerned about logging in to plinth and having full access to every service via the webinterface.
I am concerned about having to trust my whole FreedomBox security on the weakest piece of software in the chain of tools I use. 
Does this concern make sense?

**Comment by mray** [comment]
concerning: #690 deals with access to applications themselves, not handling of the credentials for application. Creation of new users should not necessarily give them the freedom to open up accounts with any installed service.

**Comment by JosephKiranBabu** [comment]
A lot has changed since this discussion. So let me brief you on the current state of things.

Our current Single Sign On solution uses an Apache module called `mod_auth_pubtkt`.
Whenever a user logs into Plinth, the Apache server generates a signed ticket that the user can produce to multiple apps running on Plinth. Suppose the user provides this ticket to Apache to request access to Syncthing, then Apache recognizes its own signed ticket and allows the user into the system.

This auth_pubtkt plugin also allows us to set tokens in the ticket, which in our case are the user groups. For example, Repro server needs the user to be admin to be able to access it. This is already working now with our current implementation of SSO.

The mechanism that worked for Repro can also be applied to other apps. The user only logs into Plinth and uses services that they have access to from Plinth. We use user groups to determine access.

I think we can currently support the following user groups using auth_pubtkt mechanism for SSO. More SSO mechanisms will be added in the future.

|App| User Group|
|-----|-----------------|
|repro|admin (change to sip later?)|
|deluge, transmission|bittorrent|
|syncthing|file-sync|
|tt-rss| news|
|ikiwiki, mediawiki|wiki|

**Comment by JoKeyser** [comment]
Hi @JosephKiranBabu , thank you for the update. Would be great to come up with a plan how fine-grained the user groups need to be, also for the other and upcoming services. I feel it would be good for users (and devs) to keep it simple with as few groups as possible (and as many as necessary). E.g. we could consider merging the bittorrent and syncthing stuff into one "file-sync"? We might start with a minimal set, and react to further wishes as they come up?

**Comment by JosephKiranBabu** [comment]
Maybe we can have one group for each category of applications. That gives a reasonable level of granularity. 
Wiki, Games, Productivity, Anonymity, Privacy, Messaging, Voip, Admin are what I can think of looking at the current applications. 
That said, we'll definitely have applications that will span multiple categories. 
I'll just create the necessary framework and implement one category in my first pull request. We'll brainstorm on the user groups and add them later.

**Comment by JosephKiranBabu** [comment]
Updated the above table. `news` is already taken. So using `feeds` instead for tt-rss. We can finalize on the names later.
